File: src/cli/commands/execute.ts
// src/cli/commands/execute.ts

import type { Command } from 'commander';
import ora from 'ora';
import chalk from 'chalk';
import { initiateExecution, pollForExecutionCompletion } from '../api_client.js';
import { handleCliError } from '../error_handler.js';

export function registerExecuteCommand(program: Command) {
  program
    .command('execute')
    .description('Execute a selected set of tests from a Test Plan.')
    .requiredOption('-p, --planId <id>', 'The Test Plan ID from the discovery phase')
    .option('-t, --tests <ids>', 'A comma-separated list of specific test component IDs to run. Leave empty to run all tests in the plan.')
    .requiredOption('--creds <profile>', 'The name of the credential profile to use')
    .action(async (options) => {
      const { planId, tests, creds: profileName } = options;

      let testsToRun: string[] | undefined = undefined;
      if (tests) {
        testsToRun = tests.split(',').map((id: string) => id.trim()).filter(Boolean);
        if (testsToRun && testsToRun.length === 0) {
          console.error(chalk.red('Error: If using the --tests flag, you must provide at least one test component ID.'));
          process.exit(1);
        }
      }

      const spinner = ora('Preparing execution...').start();
      try {

        const executionMessage = testsToRun ? `selected tests for Plan ID: ${chalk.cyan(planId)}...` : `all available tests for Plan ID: ${chalk.cyan(planId)}...`;
        spinner.text = `Initiating execution for ${executionMessage}`;

        // Pass the potentially undefined 'testsToRun' array to the API client
        await initiateExecution(planId, testsToRun, profileName);
        spinner.text = 'Execution in progress. Waiting for results...';

        const finalPlan = await pollForExecutionCompletion(planId);
        spinner.succeed(chalk.green('Execution complete!'));

        console.log('\n--- Test Execution Report ---');

        const allResults = finalPlan.planComponents.flatMap(c => c.executionResults);

        if (allResults.length === 0) {
          console.log(chalk.yellow('No tests were executed. Ensure the test IDs provided are correct.'));
          return;
        }

        let failures = 0;
        allResults.forEach(result => {
          if (result.status === 'SUCCESS') {
            console.log(`${chalk.green('PASS')} ${result.testComponentId}`);
          } else {
            failures++;
            console.log(`${chalk.red('FAIL')} ${result.testComponentId}`);
            if (result.log) {
              const indentedLog = result.log.split('\n').map(line => `  > ${line}`).join('\n');
              console.log(chalk.gray(indentedLog));
            }
          }
        });

        console.log('\n--- Summary ---');
        if (failures > 0) console.log(chalk.red(`${failures} test(s) failed.`));
        console.log(chalk.green(`${allResults.length - failures} test(s) passed.`));
        console.log(`Total tests executed: ${allResults.length}`);

      } catch (error: any) {
        spinner.fail(chalk.red('Execution failed.'));
        handleCliError(error);
      }
    });
}

File: automated-test-orchestrator-cli/cmd/execute.go
// cli-go/cmd/execute.go
package cmd

import (
	"fmt"
	"os"
	"strings"
	"time"

	"github.com/automated-test-orchestrator/cli-go/internal/client"
	"github.com/automated-test-orchestrator/cli-go/internal/display"
	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
)

// executeCmd represents the execute command
var executeCmd = &cobra.Command{
	Use:   "execute",
	Short: "Execute a selected set of tests from a Test Plan",
	Run: func(cmd *cobra.Command, args []string) {
		planID, _ := cmd.Flags().GetString("planId")
		tests, _ := cmd.Flags().GetString("tests")
		creds, _ := cmd.Flags().GetString("creds")

		var testsToRun []string
		if tests != "" {
			testsToRun = strings.Split(tests, ",")
			for i, t := range testsToRun {
				testsToRun[i] = strings.TrimSpace(t)
			}
		}

		s := spinner.New(spinner.CharSets[11], 100*time.Millisecond)
		s.Suffix = " Preparing execution..."
		s.Start()

		executionMessage := "all available tests"
		if len(testsToRun) > 0 {
			executionMessage = "selected tests"
		}
		s.Suffix = fmt.Sprintf(" Initiating execution for %s for Plan ID: %s...", executionMessage, color.CyanString(planID))

		apiClient := client.NewAPIClient(viper.GetString("api_url"))
		err := apiClient.InitiateExecution(planID, testsToRun, creds)
		if err != nil {
			s.Stop()
			color.Red("\nError: Failed to initiate execution: %v", err)
			os.Exit(1)
		}

		s.Suffix = " Execution in progress. Waiting for results..."
		finalPlan, err := apiClient.PollForExecutionCompletion(planID)
		if err != nil {
			s.Stop()
			color.Red("\nExecution failed.")
			if finalPlan != nil && finalPlan.FailureReason != nil {
				fmt.Printf("Reason: %s\n", *finalPlan.FailureReason)
			} else {
				fmt.Printf("Reason: %v\n", err)
			}
			os.Exit(1)
		}

		s.Stop()
		color.Green("✅ Execution complete!")
		display.PrintExecutionReport(finalPlan)
	},
}

func init() {
	rootCmd.AddCommand(executeCmd)
	executeCmd.Flags().StringP("planId", "p", "", "The Test Plan ID from the discovery phase (required)")
	executeCmd.Flags().StringP("tests", "t", "", "A comma-separated list of specific test component IDs to run")
	executeCmd.Flags().String("creds", "", "The name of the credential profile to use (required)")
	executeCmd.MarkFlagRequired("planId")
	executeCmd.MarkFlagRequired("creds")
}
